Mathlib/GameTheory/FairDivision/Discrete/Basic.lean

```diff
@@ -0,0 +1,144 @@
+/-
+Copyright (c) 2024 Shreyas Srinivas. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Shreyas Srinivas
+-/
+
+import Mathlib.Algebra.BigOperators.Group.Finset
+import Mathlib.Data.Real.Basic
+import Mathlib.Tactic.Linarith
+import Mathlib.Tactic.Ring.Basic
+
+
+namespace Mathlib
+
+namespace DiscreteFairDivision
+
+variable {Agents Goods : Type}
+variable [Fintype Agents] [Fintype Goods]
+variable [DecidableEq Agents]
+variable [DecidableEq Goods] -- Indivisible goods only
+
+
+
+structure Alloc where
+  assign : Agents → Finset Goods
+  disjoint : ∀ a b : Agents, a ≠ b → assign a ∩ assign b = ∅
+
+structure Val where
+  val : Agents → Finset Goods → ℝ
+  empty : ∀ a : Agents, val a ∅ = 0
+  pos : ∀ a : Agents, ∀ s : Finset Goods, val a s ≥ 0
+  mono : ∀ a : Agents, ∀ s t : Finset Goods,
+    s ⊆ t → val a s ≤ val a t
```

Comment:
I think this should be
```
abbrev Val := Agents → OrderBotHom Goods ℝ≥0
```
where `OrderBotHom` doesn't currently exist.

Reply:
Two things:
1. What is `OrderBotHom`
2. The way I have stated these definitions is basically how researchers in this field would understand these valuation functions. A lot of arguments use these individual properties both implicitly and explicitly.

Reply:
> I think this should be
> ```
> abbrev Val := Agents → OrderBotHom Goods ℝ≥0
> ```
> where `OrderBotHom` doesn't currently exist.

A third point: One can define additional properties on these valuation functions like submodularity, additivity, subadditivity etc.

Reply:
> What is OrderBotHom?

`structure OrderBotHom (A B) extends OrderHom A B, BotHom A B` or similar

> One can define additional properties on these valuation functions like submodularity, additivity, subadditivity etc.

This is perhaps an argument for writing your theorem assuming `[OrderHomClass F Goods ℝ≥0] [BotHomClass F Goods ℝ≥0]`, so that you can then apply it with `abbrev Val := Agents → F`, and still use it on other types of valuation.